contracts/modules/tokenizer/TokenizerModule.sol

```diff
@@ -1,38 +1,135 @@
 /// - Tokenize an IPA
 /// - Whitelist ERC20 Token Templates
-contract TokenizerModule is BaseModule, AccessControlled {
-    using ERC165Checker for address;
+contract TokenizerModule is
+    ITokenizerModule,
+    BaseModule,
+    AccessControlled,
+    ProtocolPausableUpgradeable,
+    UUPSUpgradeable
+{
     using Strings for *;
+    using ERC165Checker for address;
+    using IPAccountStorageOps for IIPAccount;
+
+    /// @dev Storage structure for the TokenizerModule
+    /// @param isWhitelistedTokenTemplate Mapping of token templates to their whitelisting status
+    /// @custom:storage-location erc7201:story-protocol-periphery.TokenizerModule
+    struct TokenizerModuleStorage {
+        mapping(address => bool) isWhitelistedTokenTemplate;
+    }
+
+    /// solhint-disable-next-line max-line-length
+    /// keccak256(abi.encode(uint256(keccak256("story-protocol-periphery.TokenizerModule")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 private constant TokenizerModuleStorageLocation =
+        0xef271c298b3e9574aa43cf546463b750863573b31e3d16f477ffc6f522452800;
+
+    bytes32 public constant EXPIRATION_TIME = "EXPIRATION_TIME";
 
     /// @notice Returns the protocol-wide dispute module
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
     IDisputeModule public immutable DISPUTE_MODULE;
 
+    /// @custom:oz-upgrades-unsafe-allow constructor
     constructor(
         address accessController,
         address ipAssetRegistry,
         address disputeModule
     ) AccessControlled(accessController, ipAssetRegistry) {
+        if (disputeModule == address(0)) revert Errors.TokenizerModule__ZeroDisputeModule();
+
         DISPUTE_MODULE = IDisputeModule(disputeModule);
     }
 
-    function whitelistTokenTemplate(address tokenTemplate, bool allowed) external {}
+    /// @notice Whitelists a token template
+    /// @param tokenTemplate The address of the token template
+    /// @param allowed The whitelisting status
+    function whitelistTokenTemplate(address tokenTemplate, bool allowed) external restricted {
+        if (tokenTemplate == address(0)) revert Errors.TokenizerModule__ZeroTokenTemplate();
+        if (!tokenTemplate.supportsInterface(type(IOwnableERC20).interfaceId))
+            revert Errors.TokenizerModule__UnsupportedERC20(tokenTemplate);
+
+        TokenizerModuleStorage storage $ = _getTokenizerModuleStorage();
+        $.isWhitelistedTokenTemplate[tokenTemplate] = allowed;
+
+        emit TokenTemplateWhitelisted(tokenTemplate, allowed);
+    }
+
+    /// @notice Tokenizes an IP
+    /// @param ipId The address of the IP
+    /// @param tokenTemplate The address of the token template
+    /// @param initData The initialization data for the token
+    /// @return token The address of the newly created token
+    function tokenize(
+        address ipId,
+        address tokenTemplate,
+        bytes calldata initData
+    ) external verifyPermission(ipId) returns (address token) {
```

Comment:
add one more check, that revert if the ipId has already been tokenized.

Reply:
resolved in [8efd8a2](https://github.com/storyprotocol/protocol-periphery-v1/pull/144/commits/8efd8a2d1a7e91d1a982da3735b9fa1cde96d4b4)